### PR TITLE
Allow remapping labelsets in Scorer API

### PR DESCRIPTION
**Pull request by vincentschen**
**Changes**
1. Builds API for remapping labelsets to different tasks:
```
model.score([dl_valid], remap_labels={
        "spam_task_slice:sf_link_pred": "spam_task"
})
```

2. Support "abstain_label" in `Scorer.score` (defaults to -1)

(This PR is a cleaner take on https://github.com/HazyResearch/snorkel/pull/1323— makes less assumptions about how to name labelsets)

**Test Plan**
* Added tests for new code + `tox` passes

### CODE REVIEW

File: snorkel/analysis/metrics.py [hunk 1]
```diff
@@ -53,9 +53,6 @@ def metric_score(
         msg = f"The metric you provided ({metric}) is not currently implemented."
         raise ValueError(msg)
 
-    if filter_dict is None:
```
Comment (vincentschen) [inline snorkel/analysis/metrics.py]: note: without this change, `filter_dict` was _always_ getting set, and the following if statement was not used as intended https://github.com/HazyResearch/snorkel/pull/1347/files#diff-7556b7b31429c528d90b68a0b2a3473dR62

File: snorkel/classification/snorkel_classifier.py [hunk 2]
```diff
@@ -353,13 +359,20 @@ def predict(
...
         """Calculate scores for the provided DictDataLoaders.
 
         Parameters
         ----------
         dataloaders
             A list of DictDataLoaders to calculate scores for
+        remap_labels
+            A dict specifying which labels in the dataset's Y_dict (key)
+            to remap to a new task (value)
+        kwargs
```
Comment (henryre) [inline snorkel/classification/snorkel_classifier.py]: lies

File: snorkel/classification/snorkel_classifier.py [hunk 3]
```diff
@@ -250,11 +250,13 @@ def calculate_loss(
         loss_dict = dict()
         count_dict = dict()
 
-        task_names = Y_dict.keys()
+        # Collect the available tasks for this module
+        task_names = [name for name in Y_dict.keys() if name in self.task_flows]
```
Comment (henryre) [inline snorkel/classification/snorkel_classifier.py]: helper class method? seems like we're doing this a lot

File: snorkel/classification/scorer.py [hunk 4]
```diff
@@ -36,20 +39,28 @@ def __init__(
...
     ) -> None:
 
         self.metrics: Dict[str, Callable[..., float]]
         if metrics:
             for metric in metrics:
                 if metric not in METRICS:
                     raise ValueError(f"Unrecognized metric: {metric}")
-            self.metrics = {m: partial(metric_score, metric=m) for m in metrics}
+
+            filter_dict = {"golds": [abstain_label]} if abstain_label else {}
```
Comment (henryre) [inline snorkel/classification/scorer.py]: what if `abstain_label = 0`?

File: snorkel/analysis/metrics.py (hunk 1 above)
Comment (henryre) [reply snorkel/analysis/metrics.py]: great catch. should we add a test? or was external behavior valid?

File: snorkel/classification/snorkel_classifier.py [hunk 5]
```diff
@@ -371,22 +384,36 @@ def score(self, dataloaders: List["DictDataLoader"]) -> Dict[str, float]:
...
         for dataloader in dataloaders:
             results = self.predict(dataloader, return_preds=True)
-            for task_name in results["golds"].keys():
-                metric_scores = self.scorers[task_name].score(
-                    results["golds"][task_name],
-                    results["preds"][task_name],
-                    results["probs"][task_name],
+
+            # By default, evaluate each label_name on task with corresponding task_name
+            eval_dict = {name: name for name in results["golds"].keys()}
```
Comment (henryre) [inline snorkel/classification/snorkel_classifier.py]: would `.get(task_name, task_name)` below be cleaner than forming this dict? or would this produce a better error message for a missing name?

File: snorkel/classification/snorkel_classifier.py [hunk 6]
```diff
@@ -371,22 +384,36 @@ def score(self, dataloaders: List["DictDataLoader"]) -> Dict[str, float]:
...
+            # Then, override the current results with manually specified mappings
+            eval_dict.update(remap_labels)
+
+            for label_name, task_name in eval_dict.items():
+                # Use the original gold labels, which include abstains
+                golds = (
```
Comment (henryre) [inline snorkel/classification/snorkel_classifier.py]: yuck, hard to read

```python
gold_labels = dataloader.dataset.Y_dict[label_name]  # type:ignore
gold_labels = gold_labels.numpy().squeeze()
```

File: snorkel/classification/snorkel_classifier.py [hunk 7]
```diff
@@ -371,22 +384,36 @@ def score(self, dataloaders: List["DictDataLoader"]) -> Dict[str, float]:
...
+                    dataloader.dataset.Y_dict[  # type:ignore
+                        label_name
+                    ]
+                    .numpy()
+                    .squeeze()
                 )
+                preds = results["preds"][task_name]
+                probs = results["probs"][task_name]
+                metric_scores = self.scorers[task_name].score(golds, preds, probs)
```
Comment (henryre) [inline snorkel/classification/snorkel_classifier.py]: kwargs here? have seen it elsewhere

File: snorkel/classification/scorer.py (hunk 4 above)
Comment (vincentschen) [reply snorkel/classification/scorer.py]: would work! then `filter_labels` would occur on `gold: 0`: https://github.com/HazyResearch/snorkel/blob/redux/snorkel/analysis/utils.py#L138

File: snorkel/analysis/metrics.py (hunk 1 above)
Comment (vincentschen) [reply snorkel/analysis/metrics.py]: behavior was valid—it was just confusing in the code

File: snorkel/classification/snorkel_classifier.py (hunk 3 above)
Comment (vincentschen) [reply snorkel/classification/snorkel_classifier.py]: good idea

File: snorkel/classification/snorkel_classifier.py (hunk 2 above)
Comment (vincentschen) [reply snorkel/classification/snorkel_classifier.py]: 😮 caught

File: snorkel/classification/snorkel_classifier.py (hunk 5 above)
Comment (vincentschen) [reply snorkel/classification/snorkel_classifier.py]: unfortunately, `remap_labels` is not exhaustive; it only includes the necessary overrides for certain labels. 
we also need to iterate through labels as keys, so we'd like to first populate the generic labels, and then override with those from `remap_labels`

File: snorkel/classification/snorkel_classifier.py (hunk 6 above)
Comment (vincentschen) [reply snorkel/classification/snorkel_classifier.py]: I blame `black` for formatting :). fixed!

File: snorkel/classification/snorkel_classifier.py (hunk 7 above)
Comment (vincentschen) [reply snorkel/classification/snorkel_classifier.py]: the `Scorer.score` doesn't accept kwargs: https://github.com/HazyResearch/snorkel/blob/redux/snorkel/classification/scorer.py#L53

Comment (henryre) [reply snorkel/classification/snorkel_classifier.py]: i mean `golds=golds, preds=preds...`

File: snorkel/classification/snorkel_classifier.py (hunk 5 above)
Comment (henryre) [reply snorkel/classification/snorkel_classifier.py]: ok sg

File: snorkel/classification/snorkel_classifier.py (hunk 7 above)
Comment (vincentschen) [reply snorkel/classification/snorkel_classifier.py]: got it!

File: snorkel/classification/snorkel_classifier.py [hunk 8]
```diff
@@ -330,10 +338,21 @@ def predict(
...
-                prob_dict_list[task_name].extend(prob_batch_dict[task_name])
-                gold_dict_list[task_name].extend(Y.cpu().numpy())
+            labels_to_tasks = self._get_labels_to_tasks(
+                label_names=Y_batch_dict.keys(), remap_labels=remap_labels
+            )
+
+            prob_batch_dict = self._calculate_probs(
+                X_batch_dict, labels_to_tasks.values()
+            )
+            for label_name in labels_to_tasks.keys():
```
Comment (brahmaneya) [inline snorkel/classification/snorkel_classifier.py]: for label_name in labels_to_tasks does the same thing (.keys() is implicit)

Comment (vincentschen) [reply snorkel/classification/snorkel_classifier.py]: thanks!

File: snorkel/classification/scorer.py (hunk 4 above)
Comment (henryre) [reply snorkel/classification/scorer.py]: then you need `if abstain_label is None`

Comment (vincentschen) [reply snorkel/classification/scorer.py]: good catch! fixed

**Comment by vincentschen** [comment]
Addressed! Thanks for the catch. 


On August 4, 2019 at 2:02:00 PM, Henry Ehrenberg (notifications@github.com) wrote:

@henryre requested changes on this pull request.

just blocking to make sure you see bug commented above

**Pull request merged**